File: worker/repositories/challenges.py
"""Supabase persistence for canonical Challenge records."""

from __future__ import annotations

from typing import Any, cast

from worker.challenges.models import ChallengeAIAnalysis, ChallengeNormalized
from worker.repositories.opportunities import get_service_client


def _dt(value: object) -> str | None:
    return value.isoformat() if hasattr(value, "isoformat") else None


def _row(normalized: ChallengeNormalized) -> dict[str, Any]:
    policy = normalized.ai_policy
    return {
        "source_id": normalized.source_id,
        "source_name": normalized.source_name,
        "source_type": normalized.source_type,
        "source_priority": normalized.source_priority,
        "external_id": normalized.external_id,
        "dedupe_key": normalized.dedupe_key,
        "content_hash": normalized.content_hash,
        "title": normalized.title,
        "summary": normalized.summary,
        "description": normalized.description,
        "challenge_type": normalized.challenge_type,
        "organizer": normalized.organizer,
        "host": normalized.host,
        "sponsor": normalized.sponsor,
        "start_date": _dt(normalized.start_date),
        "end_date": _dt(normalized.end_date),
        "apply_start_date": _dt(normalized.apply_start_date),
        "apply_end_date": _dt(normalized.apply_end_date),
        "result_date": _dt(normalized.result_date),
        "eligibility": normalized.eligibility,
        "eligibility_type": normalized.eligibility_type,
        "team_min": normalized.team_min,
        "team_max": normalized.team_max,
        "region": normalized.region,
        "participation_type": normalized.participation_type,
        "prize": normalized.prize,
        "total_prize_amount": normalized.total_prize_amount,
        "currency": normalized.currency,
        "prize_description": normalized.prize_description,
        "source_url": normalized.source_url,
        "application_url": normalized.application_url,
        "thumbnail_url": normalized.thumbnail_url,
        "required_documents": normalized.required_documents,
        "submission_requirements": normalized.submission_requirements,
        "technology_keywords": normalized.technology_keywords,
        "categories": normalized.categories,
        "tags": normalized.tags,
        "attachments": [item.model_dump() for item in normalized.attachments],
        "ai_policy": policy.status,
        "ai_policy_confidence": policy.confidence,
        "ai_policy_evidence": policy.evidence,
        "ai_policy_source_section": policy.source_section,
        "generative_ai_policy": normalized.generative_ai_policy,
        "ai_coding_policy": normalized.ai_coding_policy,
        "llm_policy": normalized.llm_policy,
        "ai_image_policy": normalized.ai_image_policy,
        "ai_video_policy": normalized.ai_video_policy,
        "ai_audio_policy": normalized.ai_audio_policy,
        "external_ai_api_policy": normalized.external_ai_api_policy,
        "prompt_disclosure_required": normalized.prompt_disclosure_required,
        "ai_usage_disclosure_required": normalized.ai_usage_disclosure_required,
        "analysis_status": normalized.analysis_status,
        "analysis_error": normalized.analysis_error,
        "external_api_policy": normalized.external_api_policy,
        "open_source_policy": normalized.open_source_policy,
        "copyright_policy": normalized.copyright_policy,
        "ownership_policy": normalized.ownership_policy,
        "original_text": normalized.original_text,
        "search_text": normalized.search_text,
        "status": normalized.status,
        "raw_payload": normalized.raw_payload,
        "collected_at": normalized.collected_at.isoformat(),
        "last_checked_at": normalized.last_checked_at.isoformat(),
    }


def upsert_challenge(normalized: ChallengeNormalized) -> None:
    client = get_service_client()
    response = (
        client.table("challenges")
        .select("id,source_priority,content_hash")
        .eq("dedupe_key", normalized.dedupe_key)
        .limit(1)
        .execute()
    )
    existing = cast("list[dict[str, Any]]", response.data or [])
    row = _row(normalized)
    if not existing:
        client.table("challenges").upsert(row, on_conflict="dedupe_key").execute()
        return

    current = existing[0]
    if normalized.source_priority > int(current["source_priority"]):
        client.table("challenges").update(
            {"last_checked_at": normalized.last_checked_at.isoformat()}
        ).eq("id", current["id"]).execute()
        return

    if current.get("content_hash") == normalized.content_hash:
        for key in tuple(row):
            if key.startswith("ai_") or key in {
                "analysis_status",
                "analysis_error",
                "generative_ai_policy",
                "llm_policy",
                "external_ai_api_policy",
            }:
                row.pop(key)
    client.table("challenges").update(row).eq("id", current["id"]).execute()


def get_pending_challenges(limit: int) -> list[dict[str, Any]]:
    response = (
        get_service_client()
        .table("challenges")
        .select("id,title,original_text,content_hash")
        # FAILED records retain the conservative rule-based fallback. Requeue only
        # after a content change resets analysis_status to PENDING, avoiding an
        # unbounded LLM retry/cost loop.
        .eq("analysis_status", "PENDING")
        .order("last_checked_at")
        .limit(limit)
        .execute()
    )
    return cast("list[dict[str, Any]]", response.data or [])


def update_challenge_analysis(
    challenge_id: str,
    analysis: ChallengeAIAnalysis,
    *,
    status: str,
    error: str | None,
    model: str | None,
    prompt_version: str,
) -> None:
    row: dict[str, Any] = {
        "challenge_type": analysis.challenge_type,
        "ai_policy": analysis.ai_policy.status,
        "ai_policy_confidence": analysis.ai_policy.confidence,
        "ai_policy_evidence": analysis.ai_policy.evidence,
        "ai_policy_source_section": analysis.ai_policy.source_section,
        "generative_ai_policy": analysis.generative_ai.status,
        "ai_coding_policy": analysis.ai_coding.status,
        "llm_policy": analysis.generative_ai.status,
        "ai_image_policy": analysis.ai_image.status,
        "ai_video_policy": analysis.ai_video.status,
        "ai_audio_policy": analysis.ai_audio.status,
        "external_ai_api_policy": analysis.external_ai_api.status,
        "prompt_disclosure_required": analysis.prompt_disclosure_required,
        "ai_usage_disclosure_required": analysis.ai_usage_disclosure_required,
        "analysis_status": status,
        "analysis_error": error[:1000] if error else None,
        "analysis_model": model,
        "analysis_prompt_version": prompt_version,
        "analyzed_at": None,
    }
    # PostgREST does not evaluate SQL expressions in JSON values. Use an actual timestamp.
    from datetime import UTC, datetime

    row["analyzed_at"] = datetime.now(UTC).isoformat()
    get_service_client().table("challenges").update(row).eq("id", challenge_id).execute()


def record_source_status(
    *,
    source_id: str,
    source_name: str,
    status: str,
    fetched: int,
    error: str | None,
) -> None:
    from datetime import UTC, datetime

    now = datetime.now(UTC).isoformat()
    row: dict[str, Any] = {
        "source_id": source_id,
        "source_name": source_name,
        "enabled": True,
        "last_run_at": now,
        "last_status": status,
        "fetched": fetched,
        "error": error[:1000] if error else None,
    }
    if status in {"SUCCESS", "PARTIAL_SUCCESS"}:
        row["last_success_at"] = now
    else:
        row["last_failure_at"] = now
    get_service_client().table("challenge_source_status").upsert(
        row, on_conflict="source_id"
    ).execute()

File: apps/web/src/lib/challenges.ts
import "server-only";

import { requireUser } from "@/lib/dal";
import { createClient } from "@/lib/supabase/server";

export const CHALLENGE_PAGE_SIZE = 12;

export const CHALLENGE_TYPES = [
  "CONTEST",
  "HACKATHON",
  "AI_COMPETITION",
  "DATA_COMPETITION",
  "DEV_COMPETITION",
  "IDEA_COMPETITION",
  "STARTUP_COMPETITION",
  "AWARD",
  "PUBLIC_DATA_COMPETITION",
  "OTHER",
] as const;
export const CHALLENGE_STATUSES = [
  "UPCOMING",
  "OPEN",
  "CLOSING_SOON",
  "CLOSED",
  "UNKNOWN",
] as const;
export const POLICY_STATUSES = [
  "REQUIRED",
  "ALLOWED",
  "LIMITED",
  "PROHIBITED",
  "UNKNOWN",
] as const;
export const PARTICIPATION_TYPES = ["ONLINE", "OFFLINE", "HYBRID", "UNKNOWN"] as const;

export type ChallengeType = (typeof CHALLENGE_TYPES)[number];
export type ChallengeStatus = (typeof CHALLENGE_STATUSES)[number];
export type PolicyStatus = (typeof POLICY_STATUSES)[number];
export type ParticipationType = (typeof PARTICIPATION_TYPES)[number];

export type ChallengeFilters = {
  page: number;
  q: string;
  type?: ChallengeType;
  status?: ChallengeStatus;
  ai?: PolicyStatus;
  aiCoding?: PolicyStatus;
  participation?: ParticipationType;
  entry?: "individual" | "team";
  hasPrize?: boolean;
  organizer?: string;
  technology?: string;
};

export type ChallengeSummary = {
  id: string;
  title: string;
  challenge_type: ChallengeType;
  organizer: string | null;
  apply_start_date: string | null;
  apply_end_date: string | null;
  eligibility: string | null;
  team_min: number | null;
  team_max: number | null;
  participation_type: ParticipationType;
  prize: string | null;
  total_prize_amount: number | null;
  ai_policy: PolicyStatus;
  technology_keywords: string[];
  status: ChallengeStatus;
};

export type ChallengeAttachment = {
  name: string;
  url: string | null;
  media_type: string | null;
};

export type ChallengeDetail = ChallengeSummary & {
  summary: string | null;
  description: string | null;
  host: string | null;
  sponsor: string | null;
  start_date: string | null;
  end_date: string | null;
  result_date: string | null;
  eligibility_type: string;
  region: string | null;
  prize_description: string | null;
  source_name: string;
  source_url: string;
  application_url: string | null;
  required_documents: string[];
  submission_requirements: string[];
  categories: string[];
  tags: string[];
  attachments: ChallengeAttachment[];
  ai_policy_confidence: number;
  ai_policy_evidence: string | null;
  ai_policy_source_section: string | null;
  generative_ai_policy: PolicyStatus;
  ai_coding_policy: PolicyStatus;
  llm_policy: PolicyStatus;
  ai_image_policy: PolicyStatus;
  ai_video_policy: PolicyStatus;
  ai_audio_policy: PolicyStatus;
  external_ai_api_policy: PolicyStatus;
  prompt_disclosure_required: boolean;
  ai_usage_disclosure_required: boolean;
  analysis_status: string;
  external_api_policy: string | null;
  open_source_policy: string | null;
  copyright_policy: string | null;
  ownership_policy: string | null;
  original_text: string;
};

export type ChallengePage = {
  items: ChallengeSummary[];
  total: number;
  page: number;
  pageSize: number;
};

type SearchParams = Record<string, string | string[] | undefined>;

function one(value: string | string[] | undefined): string {
  return Array.isArray(value) ? (value[0] ?? "") : (value ?? "");
}

function enumValue<T extends string>(value: string, allowed: readonly T[]): T | undefined {
  return allowed.includes(value as T) ? (value as T) : undefined;
}

export function parseChallengeFilters(params: SearchParams): ChallengeFilters {
  const rawPage = Number.parseInt(one(params.page), 10);
  const entry = one(params.entry);
  return {
    page: Number.isFinite(rawPage) && rawPage > 0 ? rawPage : 1,
    q: one(params.q).trim().slice(0, 100),
    type: enumValue(one(params.type), CHALLENGE_TYPES),
    status: enumValue(one(params.status), CHALLENGE_STATUSES),
    ai: enumValue(one(params.ai), POLICY_STATUSES),
    aiCoding: enumValue(one(params.aiCoding), POLICY_STATUSES),
    participation: enumValue(one(params.participation), PARTICIPATION_TYPES),
    entry: entry === "individual" || entry === "team" ? entry : undefined,
    hasPrize: one(params.hasPrize) === "true" || undefined,
    organizer: one(params.organizer).trim().slice(0, 100) || undefined,
    technology: one(params.technology).trim().slice(0, 50) || undefined,
  };
}

function escapeLikeTerm(term: string): string {
  return term.replace(/[%_]/g, "\\$&");
}

const SUMMARY_COLUMNS =
  "id,title,challenge_type,organizer,apply_start_date,apply_end_date,eligibility," +
  "team_min,team_max,participation_type,prize,total_prize_amount,ai_policy," +
  "technology_keywords,status";

export async function getChallenges(filters: ChallengeFilters): Promise<ChallengePage> {
  await requireUser();
  const supabase = await createClient();
  const from = (filters.page - 1) * CHALLENGE_PAGE_SIZE;
  const to = from + CHALLENGE_PAGE_SIZE - 1;

  let query = supabase
    .from("challenges")
    .select(SUMMARY_COLUMNS, { count: "exact" })
    .order("apply_end_date", { ascending: true, nullsFirst: false })
    .range(from, to);

  if (filters.q) query = query.ilike("search_text", `%${escapeLikeTerm(filters.q)}%`);
  if (filters.type) query = query.eq("challenge_type", filters.type);
  if (filters.status) query = query.eq("status", filters.status);
  if (filters.ai) query = query.eq("ai_policy", filters.ai);
  if (filters.aiCoding) query = query.eq("ai_coding_policy", filters.aiCoding);
  if (filters.participation) query = query.eq("participation_type", filters.participation);
  if (filters.entry === "individual") query = query.lte("team_min", 1);
  if (filters.entry === "team") query = query.gt("team_max", 1);
  if (filters.hasPrize) query = query.gt("total_prize_amount", 0);
  if (filters.organizer) {
    query = query.ilike("organizer", `%${escapeLikeTerm(filters.organizer)}%`);
  }
  if (filters.technology) {
    query = query.contains("technology_keywords", [filters.technology]);
  }

  const { data, error, count } = await query;
  if (error) throw new Error(`Failed to load challenges: ${error.message}`);
  return {
    items: (data ?? []) as unknown as ChallengeSummary[],
    total: count ?? 0,
    page: filters.page,
    pageSize: CHALLENGE_PAGE_SIZE,
  };
}

export async function getChallenge(id: string): Promise<ChallengeDetail | null> {
  await requireUser();
  const supabase = await createClient();
  const { data, error } = await supabase.from("challenges").select("*").eq("id", id).maybeSingle();
  if (error) throw new Error(`Failed to load challenge: ${error.message}`);
  return data as ChallengeDetail | null;
}

export type ChallengeSourceStatus = {
  source_id: string;
  source_name: string;
  enabled: boolean;
  last_run_at: string | null;
  last_success_at: string | null;
  last_failure_at: string | null;
  last_status: string | null;
  fetched: number;
  error: string | null;
};

export async function getChallengeSources(): Promise<ChallengeSourceStatus[]> {
  await requireUser();
  const supabase = await createClient();
  const { data, error } = await supabase
    .from("challenge_source_status")
    .select("source_id,source_name,enabled,last_run_at,last_success_at,last_failure_at,last_status,fetched,error")
    .order("source_name");
  if (error) throw new Error(`Failed to load challenge sources: ${error.message}`);
  return (data ?? []) as ChallengeSourceStatus[];
}

export function safeExternalUrl(value: string | null | undefined): string | null {
  if (!value) return null;
  try {
    const url = new URL(value);
    return url.protocol === "http:" || url.protocol === "https:" ? url.toString() : null;
  } catch {
    return null;
  }
}
